### PR TITLE
feature: Improve and rework mixins.

### PR DESCRIPTION
**Pull request by milesj**
## Summary

Wanted to improve mixins so that they're more versatile.

- Added more built-in mixins.
- Consumers can now define their own mixins.
- Extending mixins is now easier.
- Built-in mixins support dynamic options and are separate from common mixins.
- Removed mixins from css/less/scss as they should require a special implementation.

## Checklist

- [x] Build passes with `yarn test`.
- [x] Code is formatted with `yarn format`.
- [x] Tests have been added for my changes.
- [x] Code coverage for my change is 100%.
- [x] Documentation has been updated for my changes.

### CODE REVIEW

**Comment by milesj** [comment]
@stefhatcher Don't need to review the whole thing, but if there's CSS properties you think I should add to some of these mixins, let me know!

File: packages/compiler/templates/partials/web/palette.ejs [hunk 1]
```diff
@@ -6,16 +6,14 @@
   },
   bg: {
     base: '<%= palette.bg.base -%>',
-    disabled: '<%= palette.bg.disabled -%>',
-    focused: '<%= palette.bg.focused -%>',
-    hovered: '<%= palette.bg.hovered -%>',
-    selected: '<%= palette.bg.selected -%>',
+    <% stateOrder.forEach(state => { -%>
```
Comment (stefhatcher) [inline packages/compiler/templates/partials/web/palette.ejs:9]: niiice change

File: packages/style/src/BaseStyleSheet.ts [hunk 2]
```diff
@@ -42,8 +42,16 @@ export default abstract class BaseStyleSheet {
...
-    this.lastIndex = this.sheet.cssRules.length - 1;
+    try {
+      this.sheet.insertRule(rule, index);
+      this.lastIndex = this.sheet.cssRules.length - 1;
+    } catch {
+      // Vendor prefixed properties, pseudos, etc, that are inserted
+      // into different vendors will trigger a failure. For example,
+      // `-moz` or `-ms` being inserted into WebKit.
+      // There's no easy way around this, so let's just ignore the
+      // error so that subsequent styles are inserted.
```
Comment (stefhatcher) [inline packages/style/src/BaseStyleSheet.ts:53]: 👍🏻

File: packages/system/src/mixins/root.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    lineHeight: this.var('typography-root-line-height'),
+    textRendering: 'optimizeLegibility',
+    textSizeAdjust: '100%',
+    margin: 0,
+    padding: 0,
+    '-webkit-font-smoothing': 'antialiased',
+    '-moz-osx-font-smoothing': 'grayscale',
+  };
+
+  // Fluid typography!
```
Comment (stefhatcher) [inline packages/system/src/mixins/root.ts:20]: 🎉

**Review by stefhatcher — APPROVED** [review]

**Comment by milesj** [comment]
@stefhatcher Done for reals this time.

**Pull request merged**